pkg/apis/core/validation/validation.go

```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
 	return allErrs
 }
 
+func validateWindowsHostProcessPod(podSpec *core.PodSpec, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	containerCount := 0
+	containersWithHostProcessSetTrue := []string{}
+	containersWithHostProcessSetFalse := []string{}
+	containersWhichInheritHostProcessTrue := []string{}
+
+	var podHostProcess *bool
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.WindowsOptions != nil && podSpec.SecurityContext.WindowsOptions.HostProcess != nil {
+		podHostProcess = podSpec.SecurityContext.WindowsOptions.HostProcess
+	}
+
+	hostNetwork := false
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.HostNetwork == true {
+		hostNetwork = true
+	}
+
+	// Note: Pod-wide WindowsSecurityContextOptions are inherited by containers that do not explicitly set them.
+	for _, c := range append(podSpec.Containers, podSpec.InitContainers...) {
+		containerCount++
+		if c.SecurityContext != nil && c.SecurityContext.WindowsOptions != nil && c.SecurityContext.WindowsOptions.HostProcess != nil {
+			if *c.SecurityContext.WindowsOptions.HostProcess {
+				containersWithHostProcessSetTrue = append(containersWithHostProcessSetTrue, c.Name)
+			} else if !*c.SecurityContext.WindowsOptions.HostProcess {
+				containersWithHostProcessSetFalse = append(containersWithHostProcessSetFalse, c.Name)
+			}
+		} else if podHostProcess != nil && *podHostProcess {
+			containersWhichInheritHostProcessTrue = append(containersWhichInheritHostProcessTrue, c.Name)
+		}
+	}
+
+	hostProcessContainerCount := len(containersWithHostProcessSetTrue) + len(containersWhichInheritHostProcessTrue)
+	if hostProcessContainerCount > 0 {
+		// Fail Pod validation if feature is not enabled instead of dropping fields based on PRR reivew.
+		if !utilfeature.DefaultFeatureGate.Enabled(features.WindowsHostProcessContainers) {
+			errMsg := "Pod contains Windows HostProcess containers but feature gate 'WindowsHostProcessContainers' is not enabled"
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
+		}
+
+		// If Pod's WindowsOptions.HostProcess is set to false, no container may set WindowsOptions.HostProcess to true.
+		if podHostProcess != nil && !*podHostProcess && len(containersWithHostProcessSetTrue) > 0 {
+			errMsg := fmt.Sprintf("Pod's WindowsOptions.HostProcess is set to false but the following containers override WindowsOptions.HostProcess to true: %v", containersWithHostProcessSetTrue)
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
+		}
```

Comment:
would it be simpler to check this directly inside the container-visiting loop?

```go
for ... c in containers ...
  var containerHostProcess *bool
  if c.SecurityContext != nil && c.SecurityContext.WindowsOptions != nil {
    containerHostProcess := c.SecurityContext.WindowsOptions.HostProcess
  }

  if podHostProcess != nil && containerHostProcess != nil && *podHostProcess != *containerHostProcess {
    allErrs = append(allErrs, ... "pod hostProcess and container hostProcess must agree" error)
  }
}


```

Reply:
I was thinking of doing this but decided to group log all the errors at the end (mainly to make it more clear what all was being validated for the reader).

I think we'll still need some checks outside of the container-visiting loop but I'll can make some updates like you suggested to see if this simplifies things.